warehouse/oidc/services.py

```diff
@@ -0,0 +1,168 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+
+import redis
+import requests
+import sentry_sdk
+
+from jwt import PyJWK
+from zope.interface import implementer
+
+from warehouse.metrics.interfaces import IMetricsService
+from warehouse.oidc.interfaces import IJWKService
+from warehouse.utils import oidc
+
+
+@implementer(IJWKService)
+class JWKService:
+    def __init__(self, provider, cache_url, metrics):
+        self.provider
+        self.cache_url = cache_url
+        self.metrics = metrics
+
+        self._provider_jwk_key = f"/warehouse/oidc/jwks/{self.provider}"
+        self._provider_timeout_key = f"{self._provider_jwk_key}/timeout"
+
+    def _store_keyset(self, keys):
+        """
+        Store the given keyset for the given provider, setting the timeout key
+        in the process.
+        """
+
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            r.set(self._provider_jwk_key, json.dumps(keys))
+            r.setex(self._provider_timeout_key, 60, "placeholder")
+
+    def _get_keyset(self):
+        """
+        Return the cached keyset for the given provider, or an empty
+        keyset if no keys are currently cached.
+        """
+
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            keys = r.get(self._provider_jwk_key)
+            timeout = bool(r.exists(self._provider_timeout_key))
+            if keys is not None:
+                return (json.loads(keys), timeout)
+            else:
+                return ({}, timeout)
+
+    def _refresh_keyset(self):
+        """
+        Attempt to refresh the keyset from the OIDC provider, assuming no
+        timeout is in effect.
+
+        Returns the refreshed keyset, or the cached keyset if a timeout is
+        in effect.
+
+        Returns the cached keyset on any provider access or format errors.
+        """
+
+        # Fast path: we're in a cooldown from a previous refresh.
+        keys, timeout = self._get_keyset()
+        if timeout:
+            self.metrics.increment("warehouse.oidc.refresh_keyset.timeout")
+            return keys
+
+        oidc_url = f"{oidc.OIDC_PROVIDERS[self.provider]}/{oidc.WELL_KNOWN_OIDC_CONF}"
```

Comment:
```suggestion
        oidc_url = f"{self.token_request_url}/.well-known/openid-configuration"
```

This is the only place this is used, don't think we need it to be a constant.